src/media/inner.rs

```diff
@@ -1652,3 +1650,137 @@ fn get_or_create_source_tx<'a>(
+
+impl RtxCache {
+    fn new(max_packet_count: usize, max_packet_age: Duration) -> Self {
+        Self {
+            max_packet_count,
+            max_packet_age,
+            packet_by_seq_no: BTreeMap::new(),
+            seq_no_by_quantized_size: BTreeMap::new(),
+            last_sent_time: None,
+        }
+    }
+
+    fn cache_sent_packet(&mut self, seq_no: SeqNo, packet: Packetized, now: Instant) {
+        self.packet_by_seq_no.insert(seq_no, packet);
+        self.last_sent_time = Some(now);
+        self.remove_old_packets(now);
+    }
+
+    fn evict_acked_packet_by_seq_no(&mut self, acked_seq_no: SeqNo) {
+        self.packet_by_seq_no.remove(&acked_seq_no);
+    }
+
+    fn oldest_cached_seq_no(&self) -> Option<SeqNo> {
+        self.packet_by_seq_no.keys().next().copied()
+    }
+
+    fn last_sent_time(&self) -> Option<Instant> {
+        self.last_sent_time
+    }
+
+    fn get_cached_packet_by_seq_no(&self, seq_no: SeqNo) -> Option<&Packetized> {
+        self.packet_by_seq_no.get(&seq_no)
+    }
+
+    fn get_cached_packet_smaller_than(&self, max_size: usize) -> Option<(SeqNo, &Packetized)> {
+        let quantized_size = max_size / RTX_CACHE_SIZE_QUANTIZER;
+        let seq_no = self
+            .seq_no_by_quantized_size
+            .range(..=quantized_size)
+            .next_back()?
+            .1
+            .clone();
+        Some((seq_no, self.get_cached_packet_by_seq_no(seq_no)?))
+    }
+
+    fn remove_old_packets(&mut self, now: Instant) {
+        // TODO: test which is faster.
+        const BATCH_MODE: bool = false;
+        if BATCH_MODE {
+            if let Some(first_seq_no_thats_not_too_old) =
+                self.find_first_seq_no_thats_not_too_old(now)
+            {
+                if let Some(first_seq_no_thats_not_too_old) = first_seq_no_thats_not_too_old {
+                    self.packet_by_seq_no = self
+                        .packet_by_seq_no
+                        .split_off(&first_seq_no_thats_not_too_old);
+                } else {
+                    // They are all too old
+                    self.packet_by_seq_no.clear();
+                }
+            }
+        } else {
+            while let Some(first_seq_no_thats_too_old) = self.find_first_seq_no_thats_too_old(now) {
+                self.packet_by_seq_no.remove(&first_seq_no_thats_too_old);
+            }
+        }
+    }
+
+    fn find_first_seq_no_thats_too_old(&self, now: Instant) -> Option<SeqNo> {
+        if self.packet_by_seq_no.len() > self.max_packet_count {
+            let first_seq_no = self.packet_by_seq_no.keys().next()?;
+            // Too old because of max_packet_count.
+            return Some(first_seq_no.clone());
+        }
+        // If the max_packet_age is so old that checked_sub returns None, we shouldn't remove based on max_packet_age.
+        let min_queued_at = now.checked_sub(self.max_packet_age)?;
+
+        let (first_seq_no, first_packet) = self.packet_by_seq_no.iter().next()?;
```

Comment:
This looks a bit like what PacketizingBuffer did before we refactored it. We used to have it as a VecDeque and have SeqNo (extended seq) as the identifier. To find elements, we used `iter()` in various situations, and that's where I see similarities.

The use of `.iter()` on every send, lookup for spurious padding, calculation of queue time for the pacer, cache eviction, became a bottle neck  – it used too much CPU. This is why we changed to a ringbuffer where the `Ident` is a key into the ringbuffer that uses modulo is an instant lookup.

Iterators could work if there are less elements to consider in the first place. @xnorpx  mentioned that we could use TWCC receipts to evict from these buffers, which probably means the amount of elements would be low enough that iterating is not a problem.

Reply:
Yes, calling .iter() is a bad idea.  In my PR, I do not do any .iter().  Cached packets (for RTX) on done by SeqNo lookup in a map, or (for padding) by a bucket size.  And the FIFO for sending doesn't do any lookups: just push on one end and pop on the other.  

I agree .iter()s should be removed, especially in the hot path.  This PR accomplishes that, but by splitting the roles of FIFO and RTX cache into two separate places instead of them being done in one place.

Reply:
What are the benefits of splitting FIFO (send queue) and RTX?